File: ref/csp/templates/cpp/4-dp/3-dag.cpp
// DAG上的DP
//
// 洛谷 UVA437: The Tower of Babylon
//
// example:
//
// input:
// 1
// 10 20 30
// 2
// 6 8 10
// 5 5 5
// 7
// 1 1 1
// 2 2 2
// 3 3 3
// 4 4 4
// 5 5 5
// 6 6 6
// 7 7 7
// 5
// 31 41 59
// 26 53 58
// 97 93 23
// 84 62 64
// 33 83 27
// 0
//
// output:
// Case 1: maximum height = 40
// Case 2: maximum height = 21
// Case 3: maximum height = 28
// Case 4: maximum height = 342

#include <iostream>
#include <vector>
#include <tuple>
#include <algorithm>
#include <unordered_map>
#include <queue>

using namespace std;

using ll = long long;
using tri = tuple<int, int, int>;

int inf = 1e9;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<tri> blocks;
    vector<int> block(3);

    auto f = [&]() {
        vector<tri> vs;
        for (auto [x, y, z] : blocks) {
            vs.push_back({x, y, z});
            vs.push_back({x, z, y});
            vs.push_back({y, z, x});
        }

        tri ground = {inf, inf, 0};
        vs.push_back(ground);

        sort(vs.begin(), vs.end());
        vs.erase(unique(vs.begin(), vs.end()), vs.end());

        unordered_map<int, unordered_map<int, int>> g;
        unordered_map<int, int> indeg;

        for (int i = 0; i < vs.size(); i++) {
            for (int j = 0; j < vs.size(); j++) {
                if (i == j) {
                    continue;
                }
                tri u = vs[i];
                tri v = vs[j];
                if (get<0>(u) > get<0>(v) && get<1>(u) > get<1>(v)) {
                    g[i][j] = get<2>(v);
                    indeg[j]++;
                }
            }
        }

        queue<int> q;
        q.push(vs.size() - 1);

        vector<int> topo;

        while (!q.empty()) {
            int i = q.front();
            q.pop();

            topo.push_back(i);

            for (auto [j, w] : g[i]) {
                indeg[j]--;
                if (indeg[j] == 0) {
                    q.push(j);
                }
            }
        }

        unordered_map<int, int> dp;

        for (auto i : topo) {
            for (auto [j, w] : g[i]) {
                dp[j] = max(dp[j], dp[i] + w);
            }
        }

        int ans = 0;
        for (auto [k, v] : dp) {
            ans = max(ans, v);
        }
        return ans;
    };

    vector<int> res;
    int n;
    while (cin >> n) {
        if (n == 0) {
            break;
        }
        for (int i = 1; i <= n; i++) {
            cin >> block[0] >> block[1] >> block[2];
            sort(block.begin(), block.end());
            blocks.push_back({block[0], block[1], block[2]});
        }
        res.push_back(f());
        blocks.clear();
    }

    for (int i = 0; i < res.size(); i++) {
        cout << "Case " << i + 1 << ": maximum height = " << res[i] << endl;
    }
}

File: ref/csp/templates/python/3-problems/4-dp/3-dag.py
#!/usr/bin/env python3

# DAG上的DP
#
# 洛谷 UVA437: The Tower of Babylon
#
# example:
#
# input:
# 1
# 10 20 30
# 2
# 6 8 10
# 5 5 5
# 7
# 1 1 1
# 2 2 2
# 3 3 3
# 4 4 4
# 5 5 5
# 6 6 6
# 7 7 7
# 5
# 31 41 59
# 26 53 58
# 97 93 23
# 84 62 64
# 33 83 27
# 0
#
# output:
# Case 1: maximum height = 40
# Case 2: maximum height = 21
# Case 3: maximum height = 28
# Case 4: maximum height = 342

import sys

input = sys.stdin.readline

from collections import defaultdict
from collections import deque


def get_height(n, x, y, z):
    vertices = []
    for i in range(1, n + 1):
        # 每种立方体 -> 3个点
        vertices += [(x[i], y[i], z[i]), (x[i], z[i], y[i]), (y[i], z[i], x[i])]

    # 多源 -> 单源
    ground = (float("inf"), float("inf"), 0)
    vertices.append(ground)

    # 去重
    vertices = list(set(vertices))

    # 建图

    g = defaultdict(dict)

    indeg = defaultdict(int)  # in-degree

    for u in vertices:
        for v in vertices:
            if u == v:
                continue
            if u[0] > v[0] and u[1] > v[1]:
                g[u][v] = v[2]
                indeg[v] += 1

    # Kahn算法求拓扑排序

    q = deque()
    q.append(ground)

    topo = []

    while q:
        u = q.popleft()

        topo.append(u)

        for v in g[u]:
            indeg[v] -= 1
            if indeg[v] == 0:
                q.append(v)

    # 按拓扑序执行dp的状态转移

    dp = {}
    for u in vertices:
        dp[u] = 0

    for u in topo:
        for v in g[u]:
            dp[v] = max(dp[v], dp[u] + g[u][v])

    return max(dp.values())


res = []
while n := int(input()):
    x, y, z = ([0] * (n + 1) for _ in range(3))
    for i in range(1, n + 1):
        x[i], y[i], z[i] = sorted(map(int, input().split()))
    res.append(get_height(n, x, y, z))

for i in range(len(res)):
    print(f"Case {i + 1}: maximum height = {res[i]}")
